doc/CodeReview.md

```diff
@@ -77,7 +85,86 @@ possible responses:
 
-## 3. Reviews
+## 3. Continuous Integration
+
+Tock leans heavily on automated integration testing and as a project is
+generally willing to explore new and novel means of testing for hardware
+reliability.
+
+With exceptions for drafts or works-in-progress, generally it is expected that
+a pull request pass the full continuous integration (CI) suite before core team
+members will perform significant review labor.
+
+One frequent challenge with CI setups is replicating failures in local
+development environments. Tock goes to great lengths to mitigate this as much
+as possible. Within reason, the inability to replicate a CI test in a local
+development environment shall be considered a bug (however, it is reasonable
+that local CI requires the install of non-trivial tooling, so long as there is
+a well-documented, reliable path to set up the tooling locally).
+
+### CI Organization
+
+All CI is driven by `make` rules.
+
+Generally, there are a series of fine-grained targets that do the actual tests,
+and then a meta layer of rules that are invoked depending on context.
+
+#### The short answer: `make ci`
+This is a meta-target that runs what Tock considers the "standard developer CI".
+This is the rule that should be run locally before submitting PRs.
+It runs the quicker jobs that catch the majority of small errors.
+
+#### The complete CI setup
+
+All CI is required to support the possibility of parallel make invocation (i.e.
+`make -jN`), but is not required to handle multiple independent make processes.
+
+##### `ci-job-*`
+To the extent reasonable, individual tests are broken into small atomic units.
+All actionable `make` recipes that run tests **must** be in `ci-job-*` rules.
+These perform individual, logical actions related only to testing.
+These rules **must not** perform any one-off or setup operations.
+No automated tooling should invoke job rules directly.
+If a CI check fails, developers should be able to run the failed `ci-job-*`
+locally, although in certain cases this may require installing supporting
+tooling.
+
+##### `ci-setup-*`
+These are rules that run any required setup for jobs to succeed.
+They may install arbitrary packages or do any other significant labor.
+Many jobs may rely on the same setup target.
+To the extent possible, setup targets should cache their results to avoid
+re-execution.
+Setup targets **should** handle upgrades automatically; this may include
+automatically clearing caches or other artifacts if needed.
+Setup targets **may** handle downgrades, but developers working on experimental
+branches may be required to handle these cases manually.
+Setup targets are permitted to expect "total ownership" of the directories they
+create and manage.
+
+Setup rules may vary between runner and local environments, as they may perform
+automatic and possibly invasive (e.g. apt install) operations on runners.
+
+When run locally, setup targets **may** prompt users during install events.
```

Comment:
Mhm. This implies that a `make ci` could potentially persistently and arbitrarily change my system by doing system-wide installs and or configuration of tools, right? I think that this _may_ be helpful for some people, however there are probably a lot of people (including me) who do not want this.

On a local (non CI-infrastructure) build, I don't expect any persistent changes of my system outside of the tock working directory, or _at least_ a polite question whether the program is allowed to do so. I therefore propose changing this to a `setup targets **must** prompt users prior to system-wide persistent changes`.

Reply:
Agreed. I think in general we will avoid installing anything (as it's annoying and hard to do anyway).